File: Dynamic_programming/MATLAB_codes/can_sum_dp.m
function [result, dictionary] = can_sum_dp(target, selections, dictionary)

if target == 0
    result = true;
elseif target < 0
    result = false;
else
    keys_now = keys(dictionary);
    flag = 0;
    if min(size(keys_now))>0
        for i = 1 : max(size(keys_now))
            if keys_now(i) == target
                result = dictionary(target);
                flag = 1;
                break;
            end
        end
    end
    if flag == 0
        result = false;
        for i = 1 : size(selections, 2)
            [temp, dictionary] = can_sum_dp(target - selections(i), selections, dictionary);
            if temp == true
                result = true;
                break;
            end
        end
    end

    dictionary(target) = result;
end
end

File: Dynamic_programming/Python_Codes/can_sum_dp.py
def can_sum_dp(target, choices, dictionary):
    if dictionary.get(target) is not None:
        return dictionary.get(target), dictionary
    elif target == 0:
        return True, dictionary
    elif target < 0:
        return False, dictionary
    else:
        temp = False;
        for i in choices:
            temp, dictionary = can_sum_dp(target - i, choices, dictionary)
            if temp is True:
                break
        dictionary[target] = temp
        return temp, dictionary

target = 300
choices = [7,14]
dict_can_sum = {}
result, dict_can_sum = can_sum_dp(target, choices, dict_can_sum)

print(result)



